src/main/scala/com/github/timgent/sparkdataquality/checks/QCCheck.scala

```diff
@@ -77,6 +84,42 @@ object QCCheck {
     }
   }
 
+  trait MetricsBasedCheck[MV <: MetricValue, MC <: MetricCalculator] extends QCCheck {
+    def metricDescriptor: MetricDescriptor
+
+    def applyCheck(metric: MV): CheckResult
+
+    final def applyCheckOnMetrics(metrics: Map[MetricDescriptor, MetricValue])(implicit typeTag: ClassTag[MV]): CheckResult = {
+      val metricOfInterestOpt: Option[MetricValue] =
+        metrics.get(metricDescriptor).map(metricValue => metricValue)
+      metricOfInterestOpt match {
+        case Some(metric) =>
+          metric match {
+            case metric: MV => applyCheck(metric)
+            case _ => CheckResult(CheckStatus.Error, "Found metric of the wrong type for this check. Please report this error - this should not occur", description)
```

Comment:
Leave a todo here about potentially making the metric maps more strongly typed so there is a relationship between the abstract types in the keys and values, perhaps using - https://github.com/milessabin/shapeless/wiki/Feature-overview:-shapeless-1.2.4#heterogenous-maps